File: csharp/rational-numbers/RationalNumbers.cs
public static class RealNumberExtension
{
    public static double Expreal(this int realNumber, RationalNumber r)
    => Math.Pow(realNumber, 1.0 * r.num / r.den);
}

public struct RationalNumber
{
    public int num;
    public int den;
    public RationalNumber(int numerator, int denominator)
    {
        if(denominator == 0) { throw new ArgumentException(); }
        else if (denominator < 0)
        {
            num = -numerator;
            den = -denominator;
        }
        else
        {
            num = numerator;
            den = denominator;
        }
    }

    public static RationalNumber operator +(RationalNumber r1, RationalNumber r2)
    {
        int numerator = r1.num * r2.den + r1.den * r2.num;
        int denominator = r1.den * r2.den;
        return denominator < 0 ? 
            new RationalNumber(-numerator, -denominator).Reduce() : 
            new RationalNumber(numerator, denominator).Reduce();
    }

    public static RationalNumber operator -(RationalNumber r1, RationalNumber r2)
    => ( r1 + new RationalNumber(-r2.num, r2.den)).Reduce();

    public static RationalNumber operator *(RationalNumber r1, RationalNumber r2)
    {
        int denominator = r1.den * r2.den;
        return denominator < 0 ?
            new RationalNumber(-r1.num * r2.num, -denominator).Reduce() :
            new RationalNumber(r1.num * r2.num, denominator).Reduce();
    }

    public static RationalNumber operator /(RationalNumber r1, RationalNumber r2)
    {
        int denominator = r1.den * r2.num;
        return denominator < 0 ?
            new RationalNumber(-r1.num * r2.den, -denominator).Reduce() :
            new RationalNumber(r1.num * r2.den, denominator).Reduce();
    }

    public RationalNumber Abs()
    => new RationalNumber(Math.Abs(num), den).Reduce();

    public RationalNumber Reduce()
    {
        int a = Math.Abs(num);
        int b = den;
        while (a != 0 && b != 0)
        {
            if (a > b) a %= b;
            else b %= a;
        }
        return new (num / (a | b), den / (a | b));
    }

    public RationalNumber Exprational(int power)
    {
        if (power == 0) { return new (1, 1); }
        else if ( power < 0)
        { 
            return new RationalNumber((int)Math.Pow(den, -power), 
                                    (int)Math.Pow(num, -power)).Reduce(); 
        }
        return new RationalNumber((int)Math.Pow(num, power),
                                (int)Math.Pow(den, power)).Reduce();
    }

    public double Expreal(int baseNumber)
    => Math.Pow(baseNumber, 1.0 * this.num / this.den);
    public static implicit operator RationalNumber(int x)
    => new (x ,1);
}

File: typescript/rational-numbers/rational-numbers.ts
const getGCD = (a: number, b: number): number => (!b ? Math.abs(a) : getGCD(b, a % b));

export class Rational {
  public numerator: number;
  public denominator: number;

  constructor(numerator: number, denominator: number) {
    if (denominator === 0) {
      throw new Error('Denominator cannot be zero.');
    }
    this.numerator = numerator;
    this.denominator = denominator;
    this.reduce();
  }

  public add(other: Rational): Rational {
    const num = this.numerator * other.denominator + other.numerator * this.denominator;
    const den = this.denominator * other.denominator;
    return new Rational(num, den);
  }

  public sub(other: Rational): Rational {
    const num = this.numerator * other.denominator - other.numerator * this.denominator;
    const den = this.denominator * other.denominator;
    return new Rational(num, den);
  }

  public mul(other: Rational): Rational {
    const num = this.numerator * other.numerator;
    const den = this.denominator * other.denominator;
    return new Rational(num, den);
  }

  public div(other: Rational): Rational {
    const num = this.numerator * other.denominator;
    const den = this.denominator * other.numerator;
    return new Rational(num, den);
  }

  public abs(): Rational {
    return new Rational(Math.abs(this.numerator), Math.abs(this.denominator));
  }

  public exprational(power: number): Rational {
    if (power >= 0) {
      return new Rational(this.numerator ** power, this.denominator ** power);
    } else {
      const absPower = Math.abs(power);
      return new Rational(this.denominator ** absPower, this.numerator ** absPower);
    }
  }

  public expreal(base: number): number {
    const result = base ** (this.numerator / this.denominator);
    return Math.abs(Math.round(result) - result) < 1e-9 ? Math.round(result) : result;
  }

  public reduce(): this {
    const gcd = getGCD(this.numerator, this.denominator);
    this.numerator /= gcd;
    this.denominator /= gcd;
    if (this.denominator < 0) {
      this.numerator = -this.numerator;
      this.denominator = -this.denominator;
    }
    return this;
  }
}
